### ISSUE TITLE
Multiple B2C policies used in the one app

### ISSUE THREAD

**Issue by hills001** [AzureAD/microsoft-authentication-library-for-js#4915]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.26

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

2.3.2

### Description

Hi, 

I've been working on an Angular/B2C app based on the sample https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-angular-v2-samples/angular11-b2c-sample. 

I have created a number of custom B2C policies, but I think this same issue would occur with the built in B2C policies as well. I'm hoping to sign the user in via a B2C sign in policy, once they're logged in I want to provide links to other policies to change password and edit profile. 

I notice on line 59 of the sample app.component.ts: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/msal-angular-v2-samples/angular11-b2c-sample/src/app/app.component.ts it appears the user is logged out after completing the edit profile flow and returning to the app. Ideally I want to keep the user signed in after completing these flows, and let them continue as normal, however, if they are not logged out there will be a new set of tokens issued from the secondary policy, which causes problems with acquireTokenSilent in MSALInterceptor when calling backend APIs, as the wrong tokens can _sometimes_ be used. 

I've experimented with trying to set the account from the sign in flow as the 'active' account using setActiveAccount, however, this does not seem to work as, I'm guessing, the 'accountKey' used to reference the active account will be the same for both sets of tokens, they just have a different authority.

I can see a couple of options:

1. Use a single B2C custom policy, pass in an extraQueryParameter when calling for password change or edit profile - and use the query parameter to jump to the appropriate orchestration step. I haven't completely built this option, but it seems like it will work and the user will have only a single set of tokens after completing each flow.
2. Develop a custom version of MSALInterceptor and always pass in the tokens from the sign in flow - this doesn't seem like a great option and I can imagine dealing with multiple sets of tokens could cause other problems as well.

Is there a better to way to use multiple policies without needing to sign the user out? 

### MSAL Configuration

```javascript
export const msalConfig: Configuration = {
     auth: {
         clientId: environment.clientId, // This is the ONLY mandatory field that you need to supply.
         authority: b2cPolicies.authorities.signUpSignIn.authority, // Defaults to "https://login.microsoftonline.com/common"
         knownAuthorities: [b2cPolicies.authorityDomain], // Mark your B2C tenant's domain as trusted.
         redirectUri: '/', // Points to window.location.origin. You must register this URI on Azure portal/App Registration.
         postLogoutRedirectUri: environment.host + "#/signout"
     },
     cache: {
         cacheLocation: BrowserCacheLocation.SessionStorage, // Configures cache location. use "sessionStorage" to log user out when closing window.
         storeAuthStateInCookie: isIE, // Set this to "true" if you are having issues on IE11 or Edge
     }
}
```

### Relevant Code Snippets

```javascript
/*This is the section and comment from the sample app.component.ts which logs the user out after completing edit profile (lines 54-62)*/
/**
         * For the purpose of setting an active account for UI update, we want to consider only the auth response resulting
         * from SUSI flow. "tfp" claim in the id token tells us the policy (NOTE: legacy policies may use "acr" instead of "tfp").
         * To learn more about B2C tokens, visit https://docs.microsoft.com/en-us/azure/active-directory-b2c/tokens-overview
         */
        if (payload.idTokenClaims['tfp'] === b2cPolicies.names.editProfile) {
          window.alert('Profile has been updated successfully. \nPlease sign-in again.');
          return this.logout();
        }

/*This is the section in MSALInterceptor where the wrong tokens can sometimes be used (lines 50-62 of https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-angular/src/msal.interceptor.ts*/

 // Sets account as active account or first account
        let account: AccountInfo;
        if (!!this.authService.instance.getActiveAccount()) {
            this.authService.getLogger().verbose("Interceptor - active account selected");
            account = this.authService.instance.getActiveAccount();
        } else {
            this.authService.getLogger().verbose("Interceptor - no active account, fallback to first account");
            account = this.authService.instance.getAllAccounts()[0];
        }

        const authRequest = typeof this.msalInterceptorConfig.authRequest === "function"
            ? this.msalInterceptorConfig.authRequest(this.authService, req, { account: account })
            : { ...this.msalInterceptorConfig.authRequest, account };
```

### Identity Provider

Azure B2C Custom Policy

### Source

External (Customer)

**Comment by derisen** [comment]
@hills001 Unfortunately this has been a limitation of working with B2C with MSAL. Technically you do not need to request a token for profile edit user flow, only to take the user to [`/authorize` endpoint](https://docs.microsoft.com/en-us/azure/active-directory-b2c/authorization-code-flow#1-get-an-authorization-code) with the profile edit policy as authority, but msal-browser/msal-angular abstracts this process.

You've mentioned that you've experimented with trying to set the account from the sign in flow as the 'active' account using `setActiveAccount` -I wonder why that did not work. For the purpose of cache lookup for tokens, `accountKey` will be different since authority is different, hence the keys are distinct.

One alternative could be to [subscribe to handleRedirectObservable manually](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-angular/docs/v2-docs/redirects.md#2-subscribing-to-handleredirectobservable-manually), where you can filter out the account object that results from edit profile request and use `setActiveAccount` for the original sign-in request (since you won't be using the account & tokens obtained from profile policy anywhere else). Implementation for one such filter could be as shown [here](https://github.com/Azure-Samples/ms-identity-b2c-javascript-spa/blob/0af58afa62e8d045fe1fd6d9f2d1e49bfe5b7b41/App/authPopup.js#L27).

I'll revisit the sample see if this could be handled better, but please let us know what you find out.

**Comment by hills001** [comment]
Hi @derisen, many thanks for the reply. 

I've spent some time today working with the B2C sample and built in B2C policies to see if I could recreate the issue. I suspect there could be an issue in my own app or with my custom policies, as I haven't been able to recreate the exact issue, however, I did see some behaviour I wasn't expecting with setActiveAccount/getActiveAccount (I've set cacheLocation to SessionStorage).

I replaced the contents of 'checkAndSetActiveAccount' in app.component.ts with the following:

console.log('1=' + JSON.stringify(this.authService.instance.getActiveAccount()));
let accounts = this.authService.instance.getAllAccounts();
console.log('2=' + JSON.stringify(accounts[1]));
if (this.authService.instance.getAllAccounts().length > 0) {
  this.authService.instance.setActiveAccount(accounts[1]);
}
console.log('3=' + JSON.stringify(this.authService.instance.getActiveAccount()));
    
I also commented out lines 23 and 24 on home.component.ts to make sure the active account wasn't being set elsewhere.

On first login, I got the following in the console, which makes sense:
1=null
2=undefined
3=null

After going to the edit profile B2C policy and returning, the following was in the console:
1=null
2={homeAccountId: xxx.profile_edit...}
3={homeAccountId: xxx.sign_in...}

And on refreshing the screen I got the following:
1={homeAccountId: xxx.sign_in...}
2={homeAccountId: xxx.profile_edit...}
3={homeAccountId: xxx.sign_in...}

It seems getActiveAccount is returning a different account to the one I specified in setActiveAccount. 

After looking a little further, it seems getActiveAccount in https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/src/cache/BrowserCacheManager.ts is returning only the first account matched by getAccountInfoByFilter. 

In my case the activeAccountId which is passed to getAccountInfoByFilter matches the localAccountId value in the accounts and is the same in both accounts, but doesn't match any of the conditions in getAccountInfoByFilter so both accounts are returned. 

I'm not sure what was causing the accounts to be returned in a different order than they were created in my original app, but in that scenario I think getActiveAccount may not always point to the preferred or 'sign in' account, rather just the first matching account.
